### DISCUSSION TITLE
Allow code owners to review their own PRs

### DISCUSSION THREAD

**Discussion by frans-k** [community/community#14866] [Pull Requests]
We have these boxes checked in a branch protection rule:

- Require approvals 
- Require review from Code Owners 

I would like to check this box:

- Include administrators 

To include admins in _other_ branch protection rules, like waiting for actions, but that would also mean that admins would have to have their PRs approved by another code owner, which just isn't feasible for our setup. We also have some solo code owners.

Something along the lines of code owners approving their own PRs would solve this, which I've seen that Azure has in other discussions about this topic.

**Comment by mve-oncomfort** [comment]
I agree this is very useful to have for very small teams

**Comment by fernandomora** [comment]
Users have been asking for this since Nov 2017: https://github.community/t/do-not-require-owner-approval-if-the-pull-request-is-from-an-owner/369
We have a repository where different teams are owning different folders.
We want to explicitly approve PRs by the teams owning the folders.
But if there is a team that is composed by a single user, there is no possible way to merge if that user sends a PR.

Also, we have a automated action, that sends a PR and approves it to be merged. We needed to create 2 different bot user accounts for being able to make the workflow work, because the bot needs to belong to the CODEOWNERS team of the folder that is sending the PR, but another different bot needs to aprove since users cannot approve their own PRs

**Reply by OmgImAlexis** [reply]
Just to note that link won't work anymore since Github replaced their forums with Github discissions.

Also wayback machine won't work since it only saved the first few messages. 🙃 

https://web.archive.org/web/20210920224702/https://github.community/t/do-not-require-owner-approval-if-the-pull-request-is-from-an-owner/369/1

**Comment by tt-rkim** [comment]
+1

Has there been any movement on this? If there two codeowners on a group of files and one's on vacation, the other one is screwed.

**Reply by bdrazen** [reply]
The most absurd aspect is that if a non-codeowner creates the PR and codeowner approves, it's mergeable. If the codeowner creates the same PR and non-codeowner approves, it's not.

**Comment by teaf-wise** [comment]
Going through the same pain points described above, would be great to see some action on this

**Comment by mrr11k** [comment]
Same here. It would help us a lot, if we get this feature. 🙏

**Comment by BlueGi0** [comment]
+10000
We have Different repos owned by:
- "Devs-Team": Several Devs (Write Privileges + branch protection rules for merge on specific branches)
- "TL-Team": Technical Leader (Mantain Privileges + "Enable auto assignment" Permissions for Code Review)

So for each PR, the Techical Leader can Make de Code Review and aprobe PR's for production. But for example, in the case of an incident outside of working hours where he could make a Hot Fix, he cant't aprobe his own Pull Request.
It would be very nice to have a way to make it happen.

**Comment by pascalguttmann** [comment]
I agree. For our small team that missing functionality is a problem. Being able to approve your own PRs would be a solution.

Is there a lightweight workaround known at the moment?
Maybe it is needed to switch to a different solution like GitLab, Gitea, etc...

**Reply by LMSSonos** [reply]
My solution is to go into settings, change the branch protection and remove the required users - once the PR is passed, i add it again. Only issues, if other PRs are ready to merge but not reviewed, they will also merge. so more a hack than a solution ;-)

**Comment by gtomitsuka** [comment]
+1
Any activity on this?

**Comment by hanna-seraya** [comment]
+1 We also have a singe codeowner model, missing of this functionality lead us to always open to merge-for-everyone repo and a lot of disputes after merge. Creating "duplicated" user is not a choice for our organization, this way everyone should have duplicate

**Comment by BillPruettZen** [comment]
Please do this

**Comment by inickles** [comment]
It looks like a reasonable workaround for the model of this could be to use a Ruleset with a bypass list instead of a "classic branch protection rule". The Rulesets provide many of the same options for adding branch protections, such as "Require a pull request before merging" and "Require review from Code Owners". They also provide this bypass list feature that isn't available to the classic branch protection rules. By implementing the CODEOWNERS rule _and_ granting CODEOWNERS the ability to bypass then only the CODEOWNERS can push without review, while everyone else needs a review by them.

Individual users cannot be specified in the Bypass list, though you can get the effective equivalent if you had only one user with the repo admin role.

Those in the bypass list will see the following checkbox:

![image](https://github.com/user-attachments/assets/58e0102e-8629-4ecb-824b-39d3032eb94b)

**Reply by pascalguttmann** [reply]
@inickles Thanks for the info. That's a solution that works for our workflow. Even without the ability to bypassing the rules.
The number of `Required approvals` is set to `0`.
The settings `Require a pull request before merging`, `Require review from Code Owners` and `Dismiss stale pull request approvals when new commits are pushed` are set.

This yields bahavior as expected when working alone with as a Codeowner in a pull request.

**Reply by pascalguttmann** [reply]
@inickles 
After fiddling around and experimenting I realized another _strange_ behavior of the CODEOWNERS feature of GitHub.
Consider
- a CODEOWNER `CO` of some files and
- some collaborator `Eve`

in the following workflow:

1. `CO` does some work on CODEOWNED files, that `Eve` is not allowed to change on the protected branch without review.
2. `CO` opens a PR to the protected branch and pushes the changes.
 a. The PR is now mergeable, because `CO` is the CODEOWNER of the changed files.
3. `Eve` changes files of which `CO` is the CODEOWNER and pushes them to the existing PR.

Observed behavior:
4. `Eve` can merge the PR with the unreviewed changes to files, that are *NOT* owned by `Eve`.
Because they are part of the PR, which was created by `CO`. Even though `Eve` pushed the changes.

Expected behavior:
4. Once `Eve` pushes changes to file not owned a review of the owner `CO` is required.


I could reproduce this with rulesets and "classic branch protection".
Is this as intended, do I miss a configuration to lock `Eve` out from merging?
Or am I missing the point of `CODEOWNERS`?
(My understanding: All changes to a owned file should be made or reviewed by a codeowner of that file.)

**Reply by llegaz** [reply]
https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/managing-repository-settings/managing-teams-and-people-with-access-to-your-repository

**Reply by nyarly** [reply]
We have a requirement that all code changes require review, so setting "required approvals" to 0 isn't feasible.

The bypass list seems to work exactly the way that classic rules do, with the exception of allowing a slightly broader range of excepted users. It still spoils the ability to do auto-merge, and basically makes normal work (on a small team) into a series of compliance exceptions.

**Comment by YoshiRulz** [comment]
See also #6292.

**Comment by MichaelChirico** [comment]
My use case here:

I generate PRs for updating translations. The PR consists of a new .po file with blank translations. Translators then contribute translations and update on the original branch.

All of the actual novelty to the codebase is coming from the translators, so it makes sense that I as reviewer would review + approve their changes before merging.

**Comment by mlfarrell** [comment]
we need this.   codeowners with this behavior has led to an incredible amount of confusion among our team.  without this we basically are forced to set a single owner for a bunch of files which isn't feasible.

**Comment by DavidPriceNBS** [comment]
This is a huge blocker, I'm shocked to once more find myself on a Github discussions page which has been ignored for years, it is a trend. Every single time I hit a blocker I find a page like this which has many many people with the same issue and Github utterly ignoring it.

**Comment by christian-wolf-eks** [comment]
Did you find a solution so far?

We have the situation that there is a repo with two code owners where one (A) owns everyting in `/foo` and the other (B) everything (in `/`).

- So, when A creates a PR touching anything outside `/foo` I assume, a review by B is needed. Fine.
- If A opens a PR only touching the files outside `/foo` there seems to be no requirement for a review, so A can work solo without problems there. That is nice.
- Once either A or B touches a file in `/foo`, the review of the other is requested. If the other is absent / on holidays / ill, no PR can be merged, not even hot fixes _only_ on `/foo`. Not nice.

Is there any solution to this? I would be okay to install a GH action to review on behalf of the code owner if that us the only solution.

**Comment by dankozlowski** [comment]
Also need this!

**Comment by bdrazen** [comment]
The creation of a PR by a code owner should count as an implicit approval by a code owner. It's crazy this isn't default behaviour, let alone that this thread with hundreds of votes has gone completely ignored for coming on 4 years now.

**Reply by platinumazure** [reply]
I disagree that this should be default behavior, but I agree we should have the option to do this.

**Reply by bdrazen** [reply]
The way I see it, the author of any PR already approves of its changes by virtue of creating it. Assuming otherwise currently means PRs to files owned by a single-user team or with all but one codeowner on vacation are literally unmergeable. It also means a PR created by a non-codeowner then approved by a codeowner is mergeable, but not the other way around. This obviously makes no sense. I can't really see why anyone would want or expect this behavior, let alone preferred as default, but maybe I'm missing something.

**Comment by dark-swordsman** [comment]
I am kinda surprised that this isn't a feature. This thread has expressed that there are a lot of workarounds that kind of work, but actually cause more catastrophic edge cases, in my opinion.

At least if the required reviewers is set to 1, it could be enforced that the repo/code owner *must* at least provide some level of PR review. 

I guess I'm not really understanding the downsides to why this isn't already a feature.

I've read in other places that it's to prevent people from pushing their own changes without other eyes on the PR. However, the default behavior of a repo is to just not require code review at all... which seems more catastrophic.

For me, the code review process, even if for myself, is a way to logically separate the context. It would allow me to catch more problems with my code than otherwise.

All of this to say: I hope this feature gets added. I am surprised there are so many complex rules available, yet this seems like a pretty important one as well.

**Comment by KevinAIRegistrar** [comment]
I'm the only CODEOWNER so I literally cannot merge any pull requests as I have no-one else is a codeowner to technically no-one can approve.

WTH is this UX. How is this not fixed?

**Comment by JacobMuchow** [comment]
Please add this!!! I have lead small teams for years and commonly review/merge my own PRs. I really like the formal "submit review" button and use it on my own work regularly. I want the option to Approve or Request Changes.

**This is even more prevalent now with AI-coding,** where I am reviewing code I guided an AI to generate but haven't read in full. I like the formal marker that I self-approved it.

I understand Github's position and agree it should not be default, but there should be an override in  repo/org settings for small teams / individuals.

**Comment by igormcsouza** [comment]
My own issue is because I have many small projects that I'm the only CODEOWNER.. I have to bypass everytime.. and the protection exists because the project is open source.. which is forcing me to private the project so that I can drop the protection rule...

**Comment by VanSteve** [comment]
Please add this! I maintain my own project, but with Cursor and Claude writing the code, I find myself reviewing PRs that I have not written, but because I as the sole maintainer have the GitHub credentials, GitHub blocks me from reviewing code that I have not actually written solely because it considers my AI agents to be me. As a result, I have to have a shadow-process for reviewing and approving and marking change requests between myself and my agents, because GitHub doesn't allow me to do it properly in GitHub. Does GitHub want me to make an enterprise team just so I can add my agents as separately-licensed users? I guess that's probably it. Very disappointing.

**Comment by webern** [comment]
I'm often working by myself on side projects or with a couple of people on a side project. Now with LLM agents it would be very useful to be able to approve or request changes on "my own" pull request because, I *am* the review of the LLM's work.
